src/shared/GitLab/GitLabHostProvider.cs

```diff
@@ -0,0 +1,260 @@
+            {
+                case AuthenticationModes.Basic:
+                case AuthenticationModes.Pat:
+                    return promptResult.Credential;
+
+                case AuthenticationModes.Browser:
+                    return await GenerateOAuthCredentialAsync(input);
+
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(promptResult));
+            }
+        }
+
+        internal AuthenticationModes GetSupportedAuthenticationModes(Uri targetUri)
+        {
+            // Check for an explicit override for supported authentication modes
+            if (Context.Settings.TryGetSetting(
+                GitLabConstants.EnvironmentVariables.AuthenticationModes,
+                Constants.GitConfiguration.Credential.SectionName, GitLabConstants.GitConfiguration.Credential.AuthenticationModes,
+                out string authModesStr))
+            {
+                if (Enum.TryParse(authModesStr, true, out AuthenticationModes authModes) && authModes != AuthenticationModes.None)
+                {
+                    Context.Trace.WriteLine($"Supported authentication modes override present: {authModes}");
+                    return authModes;
+                }
+                else
+                {
+                    Context.Trace.WriteLine($"Invalid value for supported authentication modes override setting: '{authModesStr}'");
+                }
+            }
+
+            if (GitLabConstants.IsGitLabDotCom(targetUri))
+            {
+                return AuthenticationModes.All;
+            }
+
+            Context.Streams.Error.WriteLine($"Missing OAuth configuration for {targetUri.Host}, see https://github.com/GitCredentialManager/git-credential-manager/blob/main/docs/gitlab.md.");
+            // Would like to query password_authentication_enabled_for_git, but can't unless logged in https://gitlab.com/gitlab-org/gitlab/-/issues/349463
+            return AuthenticationModes.Basic | AuthenticationModes.Pat;
+        }
+
+        // <remarks>Stores OAuth tokens as a side effect</remarks>
+        public override async Task<ICredential> GetCredentialAsync(InputArguments input)
+        {
+            string service = GetServiceName(input);
+            ICredential credential = Context.CredentialStore.Get(service, input.UserName);
+            if (credential?.Account == "oauth2" && await IsOAuthTokenExpired(input.GetRemoteUri(), credential.Password))
+            {
+                Context.Trace.WriteLine("Removing expired OAuth access token...");
+                Context.CredentialStore.Remove(service, credential.Account);
+                credential = null;
+            }
+
+            if (credential != null)
+            {
+                return credential;
+            }
+
+            string refreshService = GetRefreshTokenServiceName(input);
+            string refreshToken = Context.CredentialStore.Get(service, input.UserName)?.Password;
```

Comment:
This is the incorrect service name. It should be `refreshService` or else we'll never find existing refresh tokens.

```suggestion
            string refreshToken = Context.CredentialStore.Get(refreshService, input.UserName)?.Password;
```